### PR TITLE
PLENTY-285_fix_product_url

### PR DESCRIPTION
**Pull request by TheKeymaster**

### CODE REVIEW

File: src/Product.php [hunk 1]
```diff
@@ -1038,11 +1053,9 @@ protected function processTexts(array $data)
             $this->setField('name', $this->getFromArray($texts, 'name' . $this->productNameFieldId))
                 ->setField('summary', $this->getFromArray($texts, 'shortDescription'))
                 ->setField('description', $this->getFromArray($texts, 'description'))
-                ->setField('url', $this->getProductFullUrl(
-                    $this->getFromArray($texts, 'urlPath'),
-                    $this->getFromArray($data, 'mainVariationId')
-                ))
                 ->setField('keywords', $this->getFromArray($texts, 'keywords'));
+
+            $this->path = $this->buildProductPath($this->getFromArray($texts, 'urlPath'));
```
Comment (dblaichinger) [inline src/Product.php:1058]: Use `$this->setPath()`

File: src/Product.php [hunk 2]
```diff
@@ -355,18 +369,14 @@ public function getProductFullUrl($path, $mainVariationId)
...
-        $path = '/' . ltrim($path, '/');
-        $path = rtrim($path, '/');
-
         return sprintf(
             '%s%s%s%s_%s_%s',
             $this->protocol,
             $this->getStoreUrl(),
             $prefix,
-            $path,
+            $this->path,
```
Comment (dblaichinger) [inline src/Product.php:377]: It seems `getProductFullUrl` is called before path is set, does that mean it's an empty string?

File: tests/ProductTest.php [hunk 3]
```diff
@@ -1674,6 +1675,29 @@ public function testSalesFrequencyIsProperlyExported(): void
         $this->assertSame(1337, $productMock->getField('sales_frequency'));
     }
 
+    public function testUrlIsBuiltFromFirstVariationIfMainVariationIsInactive(): void
```
Comment (dblaichinger) [inline tests/ProductTest.php:1678]: What happens when first variation is inactive as well?

Comment (dblaichinger) [reply tests/ProductTest.php:1678]: `testUrlIsBuiltFromFirstActiveVariation...`?

File: tests/ProductTest.php [hunk 4]
```diff
@@ -1674,6 +1675,29 @@ public function testSalesFrequencyIsProperlyExported(): void
...
+        $productMock->setPath('/wohnzimmer/buro');
+
+        $response = $this->getMockResponse('/pim/variations/main_variation_inactive_other_is_active.json');
+        $productMock->processVariation($response['entries'][0]);
+
+        $this->assertEmpty($productMock->getField('url'));
+
+        $productMock->processVariation($response['entries'][1]);
+
+        $this->assertSame('http://www.blub.io//wohnzimmer/buro_69_1074', $productMock->getField('url'));
```
Comment (dblaichinger) [inline tests/ProductTest.php]: Why double slash? Shouldn't the `trim` take care of that?

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: tests/ProductTest.php (hunk 4 above)
Comment (TheKeymaster) [reply tests/ProductTest.php]: That was actually an oversight of me. In the real world this does not happen, the test setup was just wrong at this point. I already fixed it before you sent the review ;).

File: tests/ProductTest.php (hunk 3 above)
Comment (TheKeymaster) [reply tests/ProductTest.php:1678]: > What happens when first variation is inactive as well?

The product won't be exported, since it has no variations.

Comment (dblaichinger) [reply tests/ProductTest.php:1678]: When main and first variation is inactive and second variation is active, it still gets the right id?

File: src/Product.php (hunk 2 above)
Comment (TheKeymaster) [reply src/Product.php:377]: No it is not. The path is set when we initially get the data from the product. The URL is now set when we go through each variation, which is done after setting the initial data from the product.

File: src/Product.php [hunk 5]
```diff
@@ -1038,11 +1053,9 @@ protected function processTexts(array $data)
             $this->setField('name', $this->getFromArray($texts, 'name' . $this->productNameFieldId))
                 ->setField('summary', $this->getFromArray($texts, 'shortDescription'))
                 ->setField('description', $this->getFromArray($texts, 'description'))
-                ->setField('url', $this->getProductFullUrl(
-                    $this->getFromArray($texts, 'urlPath'),
-                    $this->getFromArray($data, 'mainVariationId')
-                ))
                 ->setField('keywords', $this->getFromArray($texts, 'keywords'));
+
+            $this->path = $this->setPath($this->getFromArray($texts, 'urlPath'));
```
Comment (dblaichinger) [inline src/Product.php:1058]: The setter returns an object, so it shouldn't be set to `$this->path`

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: src/Product.php (hunk 5 above)
Comment (TheKeymaster) [reply src/Product.php]: Good that there are tests :p. Already fixed it ^^.

File: tests/ProductTest.php (hunk 3 above)
Comment (TheKeymaster) [reply tests/ProductTest.php]: > When main and first variation is inactive and second variation is active, it still gets the right id?

Yes of course
image
image
image

Comment (dblaichinger) [reply tests/ProductTest.php]: Then please rename test to "testUrlIsBuiltFromFirst**Active**VariationIfMainVariationIsInactive" so it's clear what the test is about.

**Review by dblaichinger — APPROVED** [review]

**Pull request merged**